### DISCUSSION TITLE
Generate Irrigation Schedule

### DISCUSSION THREAD

**Discussion by iliasmachairas** [aquacropos/aquacrop#130] [Q&A]
I would like to write some code so that AquaCrop-OSPy can suggest the irrigation schedule. I followed this tutorial regarding the Aqua Crop GUI. (https://www.youtube.com/watch?v=o5P35ogKDvw&ab_channel=FoodandAgricultureOrganizationoftheUnitedNations)
I have also studied the 4 notebooks you have attached on the documentation. I would like to to know when irrigation needs to take place. I would also like to use different water depths (in mm) over the growing season as it is shown on the video. For instance, after the 1st of May, 30mm and after the 1st of August, 50mm using a different RAW or soil moisture target.

Based on the documentation and some jupyter notebooks, I selected rrMethod=1 as the closets option to what I want to do. Irrigation is triggered if soil water content drops below a specified threshold (or four thresholds representing four major crop growth stages (emergence, canopy growth, max canopy, senescence). I have written the following code (code regarding importing the packages has been removed to keep the question short)

```
from aquacrop import AquaCropModel, Soil, Crop, InitialWaterContent, IrrigationManagement
from aquacrop.utils import prepare_weather, get_filepath
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import fmin

smts = [99]*4 # soil moisture targets [99, 99, 99, 99]
max_irr_season = 300 # 300 mm (water)
path = get_filepath('champion_climate.txt')
wdf = prepare_weather(path)
year1 = 2018
year2 = 2018
maize = Crop('Maize',planting_date='05/01') # define crop
loam = Soil('ClayLoam') # define soil
init_wc = InitialWaterContent(wc_type='Pct',value=[40]) # define initial soil water conditions

irrmngt = IrrigationManagement(irrigation_method=1,SMT=smts,MaxIrrSeason=max_irr_season) # define irrigation management
model = AquaCropModel(f'{year1}/05/01',f'{year2}/10/31',wdf,loam,maize,
irrigation_management=irrmngt,initial_water_content=init_wc)
model.run_model(till_termination=True)
```

I cannot find when and how much water (depth in mm) needs to be irrigitated. I checked model._outputs. model.irrigation_management.Schedule retrurns an array of zeros. I also tried dir(model.irrigation_management) to have a look at other methods and attributes but without any success.

Is what I am asking possible via AquaCrop-OSPy or have I misunderstood any concept?

Thanks for your time!
Kind regards,
Ilias

**Comment by Paloschi** [comment]
Yes, what you're asking is entirely possible with AquaCrop-OSPy, and you're on the right track with rrMethod=1 for your irrigation management. Your goal to adjust irrigation depths based on specific dates and soil moisture targets (SMTs) is achievable by dynamically updating the irrigation management parameters based on the growth stage or specific dates.

Your current setup with static soil moisture targets (SMTs) and a single MaxIrrSeason limit is a good starting point. However, to implement dynamic irrigation depths as per different dates (e.g., changing from 30mm after May 1st to 50mm after August 1st), you'll need to manually adjust your irrigation management strategy based on the crop's growth stage or simulation date.

AquaCrop-OSPy doesn't natively support changing irrigation depths based on specific calendar dates within a single run. Instead, you would typically adjust the IrrigationManagement parameters before starting the model run. To achieve your goal, consider simulating the growth season in parts or manually manipulating the IrrigationManagement object during the simulation. Here’s a conceptual approach:

1. Preparation Phase: Before starting the model, set up your initial IrrigationManagement with the first set of parameters (e.g., SMTs for the beginning of the season and an initial irrigation depth).

2. Running the Model in Stages: You could run the model up to a specific date (e.g., just before May 1st), then pause it, adjust the IrrigationManagement parameters (e.g., change SMTs or irrigation depth), and resume the model. Repeat this process for each date you want to adjust the irrigation parameters.

3. Manual Intervention: AquaCrop-OSPy's model runs are typically automated and continuous over the specified period. For dynamic adjustments, you might need to implement a custom loop or control structure around the model run that checks the current date and adjusts the IrrigationManagement parameters as necessary.

4. Analyzing Outputs: To find out when and how much water was applied, you should examine the output dataframes (model.Outputs.WaterProductivity and model.Outputs.Final). These outputs can give you details on the water applied on each day and the soil moisture status.

If the model.irrigation_management.Schedule returns zeros, it might be because the soil moisture never dropped below your specified thresholds, or there's an issue with how the irrigation events are being scheduled. Double-check your SMTs and ensure they are set lower than the initial soil water content to trigger irrigation.

Implementing the above approach requires a solid understanding of Python and the ability to manipulate the model's parameters dynamically. Here's a rough idea of how you could adjust the irrigation parameters mid-simulation (note that this is a conceptual approach and may need adaptation):

`
# Assuming you have a function to adjust irrigation parameters
def adjust_irrigation(model, new_smts, new_max_irr):
    model.irrigation_management.SMT = new_smts
    model.irrigation_management.MaxIrrSeason = new_max_irr

# Run model up to a specific date
model.run_model(till_termination=False, end_date='2018/04/30')

# Adjust irrigation parameters
adjust_irrigation(model, [new_smts_for_next_period], new_max_irr_for_next_period)

# Continue simulation
model.run_model(till_termination=True)
`

This approach might not be directly supported by AquaCrop-OSPy's and could require more manual control over the simulation process, including potentially directly manipulating model state variables or outputs. For a more precise and tailored solution, consulting the AquaCrop-OSPy documentation and source code.
